subprojects/core/src/integTest/groovy/org/gradle/api/ProjectConfigurationIntegrationTest.groovy

```diff
@@ -32,4 +32,128 @@ class ProjectConfigurationIntegrationTest extends AbstractIntegrationSpec {
         then:
         output.contains "the name: foobar"
     }
+
+    def "shows deprecation warning when calling Project#afterEvaluate(Closure) after the project was evaluated"() {
+        buildFile << '''
+            allprojects { p ->
+                println "[1] Adding afterEvaluate for $p.name"
+                p.afterEvaluate {
+                    println "[1] afterEvaluate $p.name"
+                }
+            }
+
+            project(':a') {
+                println "[2] Adding evaluationDependsOn"
+                evaluationDependsOn(':b')
+            }
+
+            allprojects { p ->
+                println "[3] Adding afterEvaluate for $p.name"
+                p.afterEvaluate {
+                    println "[3] afterEvaluate $p.name"
+                }
+            }
+        '''
+        settingsFile << "include 'a', 'b'"
+
+        when:
+        executer.expectDeprecationWarning()
+        run()
+
+        then:
+        outputContains("Using method Project#afterEvaluate(Closure) when the project is already evaluated has been deprecated. This will fail with an error in Gradle 7.0. Check Project#getState().getExecuted() to validate the Project wasn't already evaluated so you can defer the execution via this method.")
```

Comment:
I don't understand the advice here.  Are we suggesting that they should check `getExecuted()` before calling `afterEvaluate()` - and if they find that it has been evaluated, what should they do?  The most reasonable thing would probably be to just throw an error, I guess, but we're planning on doing that for the user anyways, so it doesn't seem like something to suggest.  Outside of that, I'm not sure what meaningful advice we could provide.  I wonder if this should just be a link to some part of the user guide where we discuss deferring configuration, like the lazy configuration chapter or something.  In other words we might say something like "many instances of deferred configuration can be handled more specifically via the Provider API - take a look at the user guide (link-to-user-guide) for more information on how to defer configuration this way"

Reply:
I think our advice here should be something more along the lines of a warning about something strange.
e.g.,
1. The configuration given is ignored because the project has already been evaluated.  To apply this configuration, remove afterEvaluate.

Reply:
I used Sterling comments and remove any mention about how it should properly be done.